### PR TITLE
Group by Backend

### PR DESCRIPTION
**Pull request by ghickman**
This adds the tables grouped by backend, and updates the docstring handling for changes in opensafely-core/cohort-extractor-v2#282

To make it a little easier to use the plugin I've made the data file location configurable.

### CODE REVIEW

File: contracts.py [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
 import pathlib
 import re
 
 from first import first
 from mkdocs.plugins import BasePlugin
 
 
-class_path_pat = re.compile(r"\n[\s]*!!!\s(.*)\n")
+backends_pat = re.compile(r"\n[\s]*!!!\sbackend:(.*)\n")
+contracts_pat = re.compile(r"\n[\s]*!!!\scontract:(.*)\n")
```
Comment (rebkwok) [inline contracts.py:11]: I feel like this is a stupid (or at least ignorant) question, but how does the docstring markdown get generated with `!!! contract:`

**Review by rebkwok — APPROVED** [review]
So now I've read this PR, and the previous one, and some more about mkdocs, and I think this all looks good, but I'll defer to anyone who knows more about all this stuff to start with

One question, but just due to my ignorance :)

File: contracts.py (hunk 1 above)
Comment (ghickman) [reply contracts.py:11]: That's a good question!  Which should probably be in the readme… 

I'll summarise here and add it to the readme in another PR.

There's couple of pieces to this…

`python -m cohortextractor generate_docs` creates a JSON file with details (name, dotted path, docstring, etc) for each of the table contracts.  It also includes the tables grouped by backend.

Mkdocs plugins follow an evented (event-driven?) pattern where plugins implement one of the `on_page_*` methods to make changes to the relevant event (so markdown for us).  That method recieves the content from each markdown file in the docs repo.  When it find somethings which matches one of the patterns above (eg a line which looks like `!!! contract:cohortextractor2.concepts.PatientDemographics`) it uses the captured pattern (`cohortextractor2.concepts.PatientDemographics`) to look up the appropriate section of the JSON file generated by cohortextractor-v2 and pull the relevant bits of info (just the name and docstring currently) for rendering into markdown.

Comment (rebkwok) [reply contracts.py:11]: OK, right, so I got all of that from my reading of the PRs and mkdocs plugin docs.  I get what's going on, but how does the line like `!!! contract:cohortextractor2.concepts.PatientDemographics` get generated?  How does that class (or how does mkdocs) know it's a contract and generate the markdown line with `!!! contract:...`?

Comment (rebkwok) [reply contracts.py:11]: Wait - docs repo.  That's it, isn't it :)

Comment (rebkwok) [reply contracts.py:11]: Got it! https://github.com/opensafely/documentation/pull/505/

**Pull request merged**